### DISCUSSION TITLE
IrScrutinizer capture does not work with Girs Nano

### DISCUSSION THREAD

**Discussion by zpuskaric** [bengtmartensson/IrScrutinizer#443] [Hardware issues] [answered]
Hi,

I am trying to capture my remotes by using Arduino Nano with Girs 1.03 backend and IrScrutinizer. Girs is uploaded to Nano and in Arduino serial console I can capture remote keys as is visible in the screenshot:

![Nano Girs 1 03 capture](https://user-images.githubusercontent.com/53512468/144311645-46410b9b-35eb-42f5-8342-1113a25af5c6.png)

After that I closed Arduino IDE and started IrScrutinizer-2.3.1-x86_64.AppImage on Ubuntu 10.04. My user is member of dialout group and selecting and configuring Girs backend (Sending hw) shows that device is connected and Girs version is 1.0.3.

![IrScrutinizer Girs Sending hw](https://user-images.githubusercontent.com/53512468/144312452-37aa27b2-a8f9-4b6f-b1b4-af284c4e65e3.png)

My problem is that capture does not work, for example when I press Test button on Capturing hw I get: "No Signal received"
When I pressed testing button yellow led lights and after reading remote key it goes back to red LED.

Is there a way to log serial connection if what I wrote does not have imediate answer?

**Comment by zpuskaric** [comment]
Ubuntu is version 20.04 (not 10.04).

**Comment by bengtmartensson** [comment]
Hi,

> Is there a way to log serial connection if what I wrote does not have imediate answer?

Select Options (in the top bar) -> verbose, and the serial communication will be logged to the console window. Please do this, perform the whole procedure,  and post the (complete) content here. Both the screenshoots look fine.

**Comment by zpuskaric** [comment]
Here is the log:

LocalSerialPortBuffered.readString: received "?GirsLite 1.0.3"
LocalSerialPortBuffered.sendString: Sent '\r'.
LocalSerialPortBuffered.readString: received "OK"
LocalSerialPortBuffered.sendString: Sent 'version\r'.
LocalSerialPortBuffered.readString: received "GirsLite 1.0.3"
LocalSerialPortBuffered.sendString: Sent 'modules\r'.
LocalSerialPortBuffered.readString: received "Base Transmit Capture 
Receive Led Parameters"
LocalSerialPortBuffered.sendString: Sent 'analyze\r'.
LocalSerialPortBuffered.readString: TIMEOUT
LocalSerialPortBuffered.sendString: Sent 'analyze\r'.
LocalSerialPortBuffered.readString: TIMEOUT


On 12/2/21 8:37 AM, Bengt Martensson wrote:
>
> Hi,
>
>     Is there a way to log serial connection if what I wrote does not
>     have imediate answer?
>
> Select Options (in the top bar) -> verbose, and the serial 
> communication will be logged to the console window. Please do this, 
> perform the whole procedure, and post the (complete) content here. 
> Both the screenshoots look fine.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/bengtmartensson/IrScrutinizer/issues/438#issuecomment-984365314>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AMYISFAZQ4UDOBB2Q62ODXTUO4O3XANCNFSM5JFOZ67Q>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
> or Android 
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>
>
-- 

Croatian Institute of Emergency Medicine <http://www.hzhm.hr>

	

Željko Puškarić / Voditelj službe za informatiku | Head of IT Sector
***@***.*** ***@***.***>/ +385 1 5496-092

Hrvatski zavod za hitnu medicinu | Croatian Institute of Emergency Medicine
Office: +385 1 5496-089 / Fax: +385 1 4677-180
Planinska 13, 10000 Zagreb, Hrvatska | Croatia
www.hzhm.hr <http://www.hzhm.hr>

**Comment by bengtmartensson** [comment]
```
LocalSerialPortBuffered.readString: received "?GirsLite 1.0.3"
LocalSerialPortBuffered.sendString: Sent '\r'.
LocalSerialPortBuffered.readString: received "OK"
LocalSerialPortBuffered.sendString: Sent 'version\r'.
LocalSerialPortBuffered.readString: received "GirsLite 1.0.3"
LocalSerialPortBuffered.sendString: Sent 'modules\r'.
LocalSerialPortBuffered.readString: received "Base Transmit Capture
Receive Led Parameters"
LocalSerialPortBuffered.sendString: Sent 'analyze\r'.
```

At this time, the Arduino starts looking for an IR signal, and the yellow LED goes on. Please have the to-be-analyzed signal sent here.

    LocalSerialPortBuffered.readString: TIMEOUT


For some reason, nothing arrives (or is recognized).

That is all I can say without further information.

**Comment by zpuskaric** [comment]
Here are the same commands and replies from Arduino IDE serial console:

Sent: \r
received: OK
Sent: version\r
received: GirsLite 1.0.3
Sent: modules\r
received: Base Transmit Capture Receive Led Parameters
Sent: analyze\r
<yellow LED turns on>
<button on remote pressed>
<red LED turns on>
received: f=36363 +867 -898 +865 -898 +1752 -898 +865 -898 +865 -899 
+865 -1788 +1750 -898 +865 -1786 +1752 -899 +865 -898 +865 -65535 +865 
-901 +865 -898 +1750 -898 +865 -898 +867 -898 +865 -1786 +1750 -899 +865 
-1788 +1750 -898 +865 -898 +865 -65535


On 12/2/21 7:56 PM, Bengt Martensson wrote:
> |LocalSerialPortBuffered.readString: received "?GirsLite 1.0.3" 
> LocalSerialPortBuffered.sendString: Sent '\r'. 
> LocalSerialPortBuffered.readString: received "OK" 
> LocalSerialPortBuffered.sendString: Sent 'version\r'. 
> LocalSerialPortBuffered.readString: received "GirsLite 1.0.3" 
> LocalSerialPortBuffered.sendString: Sent 'modules\r'. 
> LocalSerialPortBuffered.readString: received "Base Transmit Capture 
> Receive Led Parameters" LocalSerialPortBuffered.sendString: Sent 
> 'analyze\r'. |
>
> At this time, the Arduino starts looking for an IR signal, and the 
> yellow LED goes on. Please have the to-be-analyzed signal sent here.
>
> |LocalSerialPortBuffered.readString: TIMEOUT |
>
> For some reason, nothing arrives (or is recognized).
>
> That is all I can say without further information.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/bengtmartensson/IrScrutinizer/issues/438#issuecomment-984907357>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AMYISFAOGNHQL3QXAEOUSXLUO66MBANCNFSM5JFOZ67Q>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
> or Android 
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>
>
-- 

Croatian Institute of Emergency Medicine <http://www.hzhm.hr>

	

Željko Puškarić / Voditelj službe za informatiku | Head of IT Sector
***@***.*** ***@***.***>/ +385 1 5496-092

Hrvatski zavod za hitnu medicinu | Croatian Institute of Emergency Medicine
Office: +385 1 5496-089 / Fax: +385 1 4677-180
Planinska 13, 10000 Zagreb, Hrvatska | Croatia
www.hzhm.hr <http://www.hzhm.hr>

**Comment by bengtmartensson** [comment]
Please do not fullquote my text. In this context, it just adds noise. Thank you.

> Here are the same commands and replies from Arduino IDE serial console:
> ...
> received: f=36363 +867 -898 +865 -898 +1752 -898 +865 -898 +865 -899
+865 -1788 +1750 -898 +865 -1786 +1752 -899 +865 -898 +865 -65535 +865
-901 +865 -898 +1750 -898 +865 -898 +867 -898 +865 -1786 +1750 -899 +865 
-1788 +1750 -898 +865 -898 +865 -65535

This is perfect. So the hardware works.

I have just tried with Ubunto 20.04.01 and IrScrutinizer 2.3.1 AppImage and I cannot reproduce. Everything work fine. So I do not know how I can help you...??

**Comment by zpuskaric** [comment]
Hi

I am trying to debug the problem and for that matter i installed serial 
port monitoring program called interceptty. It attaches to real serial 
port and creates virtual serial port to which a program can be attached 
(in my case real port is /dev/ttyUSB0 and virtual is /dev/ttyUSB3). I 
can connect to port with minicom and everything works OK (I can send 
commands and am getting expected replies).

When I start IrScrutinizer it does see virtual port (/dev/ttyUSB3) but 
when I try to connect to it (<Open> button in Girs Client Sending hw) it 
returne message: "NonExistingPortException: /dev/ttyUSB3"

Is there a way around it? Perhaps some other serial port monitoring 
program that works with IrScrutinizer?

P.S. interceptty command line is:

interceptty -s 'ispeed 115200 ospeed 115200' /dev/ttyUSB0 /dev/ttyUSB3

**Comment by bengtmartensson** [comment]
> Is there a way around it? Perhaps some other serial port monitoring
program that works with IrScrutinizer?

With the verbose option, the serial communication is echoed to the console, so there is no real need for a port-sniffer.

Please flash "my" 1.0.3 using e.g. 
 https://github.com/bengtmartensson/AGirs/releases/download/Version-1.0.3/GirsLite-1.0.3-nano-flasher.sh

Please also try the "Generic Binary" as per http://www.harctoolbox.org/IrScrutinizer.html#Generic+Binary

**Comment by bengtmartensson** [comment]
Have you made any further progress on this?

Again, you are describing a sort-of "this-cannot-happen" situation (bidirectional serial communication first works (version number and modules), then it suddenly does not work). And I cannot reproduce...

**Comment by zpuskaric** [comment]
Hi,

I intend to try it on another computer but have been busy lately. Will
let you know the outcome.

**Comment by zpuskaric** [comment]
Hi,

I tried it on another machine but with the same result. Could it be a kernel version issue (I am using rather recent kernel versions, 5.8 on one machine and 5.13 on another)?
Is there some way to check what's going on on low level?

**Comment by zpuskaric** [comment]
Hi,
I played with Arduino Nano, Agirs and python and found possible reason why IrScrutinizer would not be able to decode captured signal.
Replies from Agirs to commands like version and modules end with -CR--LF-  but replies to capture and receive do not.

Examples:

command: version-CR -
reply: GirsLite 1.0.3-CR--LF-

command: modules-CR-
reply: Base Transmit Capture Receive Led Parameters -CR--LF-

but for capture
command: capture-CR-
reply: f=38461 +9030 -4527 +568 -569 +566 -569 +568 -1677 +568 -570 +566 -569 +566 -569 +568 -569 +568 -567 +566 -1677 +568 -1677 +568 -569 +566 -1677 +568 -1675 +568 -1677 +568 -1675 +568 -1673 +568 -1677 +568 -1675 +568 -1677 +566 -567 +568 -1675 +568 -567 +566 -569 +565 -567 +566 -567 +566 -569 +568 -567 +568 -1674 +568 -567 +568 -1679 +568 -1674 +568 -1676 +566 -40086 +9021 -2256 +568 -65535

and receive
command: receive-CR-
reply: +8900 -4500 +550 -550 +550 -600 +550 -1650 +550 -600 +550 -550 +550 -600 +550 -550 +550 -600 +550 -1650 +550 -1650 +550 -600 +500 -1700 +550 -1650 +550 -1700 +550 -1650 +550 -1700 +550 -550 +550 -600 +550 -550 +550 -1650 +550 -1700 +550 -550 +550 -600 +500 -600 +550 -1650 +550 -1700 +550 -1650 +550 -600 +500 -600 +550 -1700 +500 -1700 +550 -1650 +550 -39700 +8900 -2250 +550 -50100

So Agirs does not output line endings of any kind as a reply to capture and receive commands. If software like IrScrutinizer tries to read such data from serial port it can happen that it times out because of the lack of line ending character.

**Comment by bengtmartensson** [comment]
> So Agirs does not output line endings of any kind as a reply to capture and receive commands.

No, this is not true. All output is made with (or ended with) `Stream.println(...)`, (which happens to use CR LF, see https://www.arduino.cc/reference/en/language/functions/communication/serial/println/ ),  For example `capture`in `Girr.cpp`:

```
    if (!irWidget->isEmpty()) {
        // Trying to decode the capture does not make sense,
        // that is what "receive" is for.
        irWidget->dump(stream);
        stream.println();
    } else
        stream.println(F(timeoutString));
```

**Comment by zpuskaric** [comment, marked as answer]
Hi,

I got it working. The problem was that I installed AGirs through Arduino Library Manager. That version does not have 
`stream.println(); `
command in the right place (diff of Arduino version and GitHub version is attached). I had to remove AGirs and install it from zip file that I downloaded from GitHub. After uploading that sketch to Arduino Nano everything works fine from IrScrutinizer as well as from my Python program.
Thanks a lot for your help
[diffs.zip](https://github.com/bengtmartensson/IrScrutinizer/files/8442941/diffs.zip)
.

**Comment by bengtmartensson** [comment]
Congratulations!

Is it correct that you are using Infrared4Arduino version 1.2.2 (see ~/Arduino/libraries/Infrared/library.properties); using the library manager version 1.0.3 the problem occurs; with the current GitHub version (which is stupidly enough also called 1.0.3, blush...) it works? Please verify; I want to fix this.

**Comment by zpuskaric** [comment]
Yes, Infrared4Arduino version is indeed 1.2.2. and problem is as you described (library manager version does not work, GitHub version does work).

**Comment by bengtmartensson** [comment]
Thank you again for finding this problem! I have set the version of the current Github sources to 1.0.4 (https://github.com/bengtmartensson/AGirs/issues/64). It should be in the Arduino library manager within a few hours or so.

Let me know if you find further issues.

**Reply by zpuskaric** [reply]
Hi,
I tried and installing AGirs library through Arduino Library Manager worked OK, it offered version 1.0.4 and after installation everything works OK.
Thanks a lot one more time.

**Reply by bengtmartensson** [reply]
Glad that I could help.

Actually, I am the one who should be thankful. I  changed the API for Infrared4Arduino, and was working on changing AGirs accordingly, but somehow I got interrupted and forgot about it before checking in. Embarrassing...
